hail/src/main/scala/is/hail/utils/package.scala

```diff
@@ -1024,26 +1031,82 @@ package object utils
 
     var err = init
     val buffer = new mutable.ArrayBuffer[(A, Int)](tasks.length)
+    val completer = new ExecutorCompletionService[(A, Int)](executor)
+
+    tasks.foreach { case (t, k) => completer.submit(() => t() -> k) }
+    tasks.foreach { _ =>
+      try buffer += completer.take().get()
+      catch {
+        case e: ExecutionException =>
+          err = accum(err, e.getCause)
+        case NonFatal(ex) =>
+          err = accum(err, ex)
+      }
+    }
 
-    tasks
-      .map { case (t, k) => (executor.submit(() => Try(t())), k) }
-      .foreach { case (f, k) =>
-        f.get() match {
-          case Success(v) =>
-            buffer += ((v, k))
+    (err, buffer.sortBy(_._2))
+  }
 
-          case Failure(t) =>
-            err = accum(err, (t, k))
-        }
+  def runAllKeepFirstError[A](executor: ExecutorService)
+    : IndexedSeq[(() => A, Int)] => (Option[Throwable], IndexedSeq[(A, Int)]) =
+    runAll[Option, A](executor) { case (opt, e) => opt.orElse(Some(e)) }(None)
+}
+
+class CancellingExecutorService(delegate: ExecutorService) extends AbstractExecutorService {
```

Comment:
This `ExecutorService` cancels all tasks should one fail. 
Note `newTaskFor` is not thread safe - I'm assuming all work will be submitted from one thread (as is the case now)